packages/core/src/stateUtils.ts

```diff
@@ -1491,15 +1530,27 @@ interface BuiltinAction {
   execute: (actorScope: AnyActorScope, params: unknown) => void;
 }
 
-export let executingCustomAction:
-  | ActionFunction<any, any, any, any, any, any, any, any, any>
-  | false = false;
+export interface ExecutableAction {
+  /**
+   * The action type
+   */
+  type: string;
+  // TODO: don't know if we need this
+  _info: ActionArgs<MachineContext, EventObject, EventObject>;
+  params: NonReducibleUnknown;
+  /**
+   * Executes the action with the params (and extra action info, like context/event, for now)
+   */
+  execute: () => void;
+}
+
+export type ActionExecutor = (actionToExecute: ExecutableAction) => void;
 
 function resolveAndExecuteActionsWithContext(
   currentSnapshot: AnyMachineSnapshot,
   event: AnyEventObject,
   actorScope: AnyActorScope,
-  actions: UnknownAction[],
+  actions: UnknownActionObject[],
```

Comment:
this type is incorrect, enqueued actions can reach this as functions :P

Reply:
It is now correct; only action objects reach this now

Reply:
it's still not correct, you can easily test it by inserting a code like this here:
```ts
if (actions.some(a => typeof a === 'function') {
  throw new Error('oops')
}
```

A ton of tests will fail. The function body here checks `typeof action === 'function'` and performs specific logic related to that but TS doesn't see that it's even an acceptable solution:
```ts
// within the loop
    if (typeof action === 'function') {
      action // never
    }
```

Reply:
The reason this is `UnknownActionObject` is that it expects all the actions to be action objects, whether built-in (which are functions too) or not.

A more accurate check is this:

```ts
  if (actions.some((a) => !('type' in a))) {
    throw new Error('oops');
  }
```

And all tests still pass with that

Reply:
This is a different test and it tests a different thing. I'm saying that this type claims that functions can't be received by this function but they totally can be - and the throw-based snippet I proposed proves that they can be received here at runtime. That, in turn, means that this type annotation is not correct - or that parts of the implementation were not adjusted to make this annotation correct

Reply:
This has been reverted